### ISSUE TITLE
[BUG]: `let` initialization is not detected in a `for` loop

### ISSUE THREAD

**Issue by isuckatcs** [modularml/mojo#1414]
### Bug description

In the following snippet an error message is reported, when `x` is mutated:
```python
fn main():
    let x: Int
    if len(sys.argv()) > 10:
        x = 1
    x = 2
```
```
error: invalid mutation of immutable value 'x.value'
    x = 2
        ^
note: 'x' declared here
    let x: Int
    ^
mojo: error: failed to run the pass manager
```
The same however is not detected if the assignment is inside a `for` loop, and the following snippet compiles.
```python
fn main():
    let x: Int
    for i in range(0, 3):
        x = 1
    x = 2
```
In neither a `for` loop, nor an `if` statement the compiler can always guarantee, whether the body is entered or not (e.g.: `for in range(0, len(sys.argv()))`), so I assume the reasoning should be the same in both cases.

Currently in the first case the compiler assumes that the body of the condition is entered, and `x` is initialized there, so `x = 2` is a mutation and an error. In the second example the compiler doesn't assume that the loop will be entered (even though the boundaries are known in compile time) and `x = 2` is not flagged as a mutation. 

By the way, doesn't every iteration after the first one mutate the value? Shouldn't mojo error out _inside_ the loop?

### Steps to reproduce

- compile/run the snippets in the description with mojo

### System information

```shell
- Ubuntu 22.04.3 LTS (WSL 2)
- mojo 0.6.0 (d55c0025)
- modular 0.2.2 (95d42445)
```

**Comment by lattner** [comment]
Great catch, I agree, this is a bug.  I'm raising a discussion about what we want to do here, we may end up going with this suggestion: https://github.com/modularml/mojo/issues/1205

**Comment by duckki** [comment]
I like late initialization so I can drop unnecessary initializations before if-statements.
I think it's reasonable to prevent loops from initializing an immutable local variables (regardless whether it's guaranteed to enter only once).

A similar situation with a closure, which is currently not allowed and an acceptable limitation:
```mojo
fn main():
    let v: Int

    fn init_v():
        v = 1 #  error: expression must be mutable in assignment

   init_v()
```

**Comment by lattner** [comment]
We are pursuing the path of removing let's declarations, which resolve this class of issue.  There is another issue (#1257) where we don't check that vars are fully initialized when closures are formed, but we'll track fixing that over in that issue.
